File: src/audioWfc/wfc/constraints/MelodyStartsOnNoteHardConstraint.java
package audioWfc.wfc.constraints;

import audioWfc.wfc.HigherValues;
import audioWfc.musicTheory.OctavedNote;
import audioWfc.wfc.Tile;
import audioWfc.wfc.constraints.concepts.HardConstraint;
import audioWfc.wfc.grabbers.Grabber;
import audioWfc.musicTheory.Note;

public class MelodyStartsOnNoteHardConstraint implements HardConstraint<OctavedNote> {
    private Grabber<Note> grabber;

    public MelodyStartsOnNoteHardConstraint(Grabber<Note> grabber){
        this.grabber = grabber;
    }

    @Override
    public boolean check(Tile<OctavedNote> tile, HigherValues higherValues) {
        if(tile.getPosition() != 0) return true;
        return tile.getValue().getNote().equals(grabber.grab(higherValues));
    }

}

File: src/wfc/constraints/MelodyStartsOnNoteHardConstraint.ts
import { OctavedNote, Note } from "../../music_theory/Note"
import { Grabber, NoteGrabberIR } from "../Grabber"
import { HigherValues } from "../HigherValues"
import { Tile } from "../Tile"
import { HardConstraint } from "./concepts/Constraint"
import { noteConstraintTypeToName } from "./constraintUtils"

export const MelodyStartsOnNoteHardConstraintInit = {
	type: "MelodyStartsOnNoteHardConstraint" as const,
	noteGrabber: "RootOfChordGrabber" as NoteGrabberIR,
	validByDefault: true as const,
}

export type MelodyStartsOnNoteHardConstraintIR =
	typeof MelodyStartsOnNoteHardConstraintInit

export class MelodyStartsOnNoteHardConstraint
	implements HardConstraint<OctavedNote>
{
	name = noteConstraintTypeToName.get(
		MelodyStartsOnNoteHardConstraintInit.type,
	)!.name as string
	constructor(private grabber: Grabber<Note | undefined>) {}

	check(tile: Tile<OctavedNote>, higherValues: HigherValues): boolean {
		if (tile.getPosition() != 0) return true
		const note = tile.getValue()
		const startNote = this.grabber(higherValues)
		return startNote === undefined || note.getNote() == startNote
	}
}
